File: Android_Hospital_Management/Modern_Hospital_Management_System-Backend/src/main/java/emranhss/com/Modern_Hospital_Management_System/service/StockService.java
package emranhss.com.Modern_Hospital_Management_System.service;

import emranhss.com.Modern_Hospital_Management_System.dto.request.MedicineStockRequest;
import emranhss.com.Modern_Hospital_Management_System.dto.request.StockAdjustmentRequest;
import emranhss.com.Modern_Hospital_Management_System.dto.response.MedicineStockResponse;
import emranhss.com.Modern_Hospital_Management_System.dto.response.StockHistoryResponse;

import java.util.List;

public interface StockService {
    MedicineStockResponse addStock(MedicineStockRequest request);
    MedicineStockResponse updateStock(Long id, MedicineStockRequest request);
    MedicineStockResponse getById(Long id);
    List<MedicineStockResponse> getAll();
    List<MedicineStockResponse> getAvailable();
    List<MedicineStockResponse> search(String keyword);
    List<MedicineStockResponse> getLowStock();
    List<MedicineStockResponse> getExpired();
    List<MedicineStockResponse> getExpiringSoon(int days);
    List<MedicineStockResponse> getByBatch(String batch);
    List<MedicineStockResponse> getBySupplier(Long supplierId);
    MedicineStockResponse getByBarcode(String barcode);
    MedicineStockResponse adjustStock(StockAdjustmentRequest request);
    List<StockHistoryResponse> getStockHistory(Long stockId);
    void decreaseStock(Long stockId, int quantity);
    void increaseStock(Long stockId, int quantity, String reason, String performedBy);
    void delete(Long id);
}

File: SPRING_BOOT_Hospital_Management/Hospital_Management_system_angular/src/app/services/stock.service.ts
import { Injectable } from '@angular/core';
import { environment } from '../../environments/environment';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { MedicineStockModel } from '../models/medicine-stock.model';
import { StockAdjustmentModel, StockHistoryModel } from '../models/stock-adjustment.model';

@Injectable({ providedIn: 'root' })
export class StockService {
  private api = environment.apiUrl + 'stock';
  constructor(private http: HttpClient) {}
  addStock(data: MedicineStockModel): Observable<MedicineStockModel> { return this.http.post<MedicineStockModel>(this.api, data); }
  updateStock(id: number, data: MedicineStockModel): Observable<MedicineStockModel> { return this.http.put<MedicineStockModel>(`${this.api}/${id}`, data); }
  getById(id: number): Observable<MedicineStockModel> { return this.http.get<MedicineStockModel>(`${this.api}/${id}`); }
  getAll(): Observable<MedicineStockModel[]> { return this.http.get<MedicineStockModel[]>(this.api); }
  getAvailable(): Observable<MedicineStockModel[]> { return this.http.get<MedicineStockModel[]>(`${this.api}/available`); }
  search(keyword: string): Observable<MedicineStockModel[]> { return this.http.get<MedicineStockModel[]>(`${this.api}/search?keyword=${encodeURIComponent(keyword)}`); }
  getLowStock(): Observable<MedicineStockModel[]> { return this.http.get<MedicineStockModel[]>(`${this.api}/low-stock`); }
  getExpired(): Observable<MedicineStockModel[]> { return this.http.get<MedicineStockModel[]>(`${this.api}/expired`); }
  getExpiringSoon(days: number = 30): Observable<MedicineStockModel[]> { return this.http.get<MedicineStockModel[]>(`${this.api}/expiring-soon?days=${days}`); }
  getByBatch(batch: string): Observable<MedicineStockModel[]> { return this.http.get<MedicineStockModel[]>(`${this.api}/batch/${batch}`); }
  getBySupplier(supplierId: number): Observable<MedicineStockModel[]> { return this.http.get<MedicineStockModel[]>(`${this.api}/supplier/${supplierId}`); }
  getByBarcode(barcode: string): Observable<MedicineStockModel> { return this.http.get<MedicineStockModel>(`${this.api}/barcode/${barcode}`); }
  adjustStock(data: StockAdjustmentModel): Observable<MedicineStockModel> { return this.http.post<MedicineStockModel>(`${this.api}/adjust`, data); }
  getStockHistory(stockId: number): Observable<StockHistoryModel[]> { return this.http.get<StockHistoryModel[]>(`${this.api}/${stockId}/history`); }
  delete(id: number): Observable<any> { return this.http.delete(`${this.api}/${id}`); }
}
